tuta-sdk/rust/src/rest_error.rs

```diff
@@ -0,0 +1,432 @@
+        use DomainFailureReason::*;
+        match s {
+            "domain.contact_form_active" => Ok(ContactFormActive),
+            "domain.not_a_subdomain" => Ok(NotASubdomain),
+            "domain.invalid" => Ok(Invalid),
+            "domain.invalid_cname" => Ok(InvalidCDomain),
+            "domain.exists" => Ok(Exists),
+            _ => Err(ParseFailureError)
+        }
+    }
+}
+
+/// The failed preconditions of a REST request to attempt setting a custom domain
+/// for an email address
+#[derive(Error, Debug, uniffi::Enum, Eq, PartialEq)]
+pub enum CustomDomainFailureReason {
+    #[error("LimitReached")]
+    LimitReached,
+    #[error("DomainInUse")]
+    DomainInUse,
+}
+
+impl FromStr for CustomDomainFailureReason {
+    type Err = ParseFailureError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        use CustomDomainFailureReason::*;
+        match s {
+            "customdomainservice.limit_reached" => Ok(LimitReached),
+            "customdomainservice.domain_in_use" => Ok(DomainInUse),
+            _ => Err(ParseFailureError)
+        }
+    }
+}
+
+/// The failed preconditions of unsuccessful template group operations
+#[derive(Error, Debug, uniffi::Enum, Eq, PartialEq)]
+pub enum TemplateGroupFailureReason {
+    #[error("BusinessFeatureRequired")]
+    BusinessFeatureRequired,
+    #[error("UnlimitedRequired")]
+    UnlimitedRequired,
+}
+
+impl FromStr for TemplateGroupFailureReason {
+    type Err = ParseFailureError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        use TemplateGroupFailureReason::*;
+        match s {
+            "templategroup.business_feature_required" => Ok(BusinessFeatureRequired),
+            "templategroup.unlimited_required" => Ok(UnlimitedRequired),
+            _ => Err(ParseFailureError)
+        }
+    }
+}
+
+/// The failed preconditions of unsuccessful usage test operations
+#[derive(Error, Debug, uniffi::Enum, Eq, PartialEq)]
+pub enum UsageTestFailureReason {
+    #[error("InvalidState")]
+    InvalidState,
+    #[error("InvalidRestart")]
+    InvalidRestart,
+    #[error("InvalidStage")]
+    InvalidStage,
+    #[error("InvalidStageSkip")]
+    InvalidStageSkip,
+    #[error("InvalidStageRepetition")]
+    InvalidStageRepetition,
+}
+
+impl FromStr for UsageTestFailureReason {
+    type Err = ParseFailureError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        use UsageTestFailureReason::*;
+        match s {
+            "invalid_state" => Ok(InvalidState),
+            "invalid_restart" => Ok(InvalidRestart),
+            "invalid_stage" => Ok(InvalidStage),
+            "invalid_stage_skip" => Ok(InvalidStageSkip),
+            "invalid_stage_repetition" => Ok(InvalidStageSkip),
+            _ => Err(ParseFailureError)
+        }
+    }
+}
+
+/// The possible failed preconditions when unsuccessfully performing an operation on the backend
+#[derive(Error, Debug, uniffi::Enum, Eq, PartialEq)]
+pub enum PreconditionFailedReason {
```

Comment:
I am not sure it makes sense to collect all precondition errors into a single enum.

As you can see the error depends on the service we are calling, we cannot ever have `UnsubscribeFailure` and `UsageTestFailure` for the same call.

Additionally, there are always possible preconditions that the client is not aware of yet as they don't benefit from our automated migration system.

I thought it would be out of scope of this PR to try and map all of these errors because we won't expose the methods (yet) to call the services.

Reply:
> I am not sure it makes sense to collect all precondition errors into a single enum.
> 
> As you can see the error depends on the service we are calling, we cannot ever have `UnsubscribeFailure` and `UsageTestFailure` for the same call.

My thinking was type-wise. All of the `FailureReasons` are a type of `PreconditionFailedReason` so if a function needs to handle both a `UnsubscribeFailureReason` and a `BookingFailureReason` (which is done by `handleSwitchAccountPreconditionFailed()` in `SwitchSubscriptionDialog.ts`), we can declare the type as a `PreconditionFailedReason`. Otherwise if it just need to handle a particular type of `PreconditionFailedReason`, we can declare the type as that specific `PreconditionFailedReason` (say `CustomDomainFailureReason`).

> Additionally, there are always possible preconditions that the client is not aware of yet as they don't benefit from our automated migration system.

Fair enough. It sounds like `PreconditionFailedReason` needs to be generated in the future.

> I thought it would be out of scope of this PR to try and map all of these errors because we won't expose the methods (yet) to call the services.

That is true; It is out of scope to handle precondition failures. I got a bit too carried away when porting the TypeScript version of `rest_error`.

Reply:
I will not insist on enum for all preconditions but I want to elaborate a little bit.

First of all even if we generate precondition reasons the server might be never than the client. We have to add "unknown" case. This is a common practice when working with backward-compatible APIs.

Second, why I am against enumerating all precondition reasons.

I see API as structured in layers. There's a "core" layer that just knows about general tutadb machinery: apps, entities, services. It also has generic API conventions around HTTP (the list of rest errors, suspension, etc.). These are the things that are not bound to a specific app nor a specific service.

Precondition reasons are a layer on top of that. They are always bound to a specific method of a specific service. It would be great if it was part of the generated API but it doesn't change that it's tied to a specific call of a specific service. There's basically no use case no enumerate all possible preconditions and it ties together the services and apps that should be independent from each other.